File: server/src/main/kotlin/io/tuzzy/portal/api/ApiSpec.kt
package io.tuzzy.portal.api

import io.javalin.http.Context
import io.tuzzy.portal.domain.SpecStatus

data class ApiSpec(
    val apiName: String? = null,
    val specVersion: String = "v1",
    val status: SpecStatus,
    val specUrl: String? = null,
    var spec: Map<String, Any>? = null
) : HalResource() {
    fun withHal(ctx: Context): ApiSpec {
        if (apiName == null) throw IllegalStateException("Api name is not initiated, cannot build hal")

        links.addAll {
            HalBuilder(ctx)
                .toSpec("self", apiName, specVersion)
                .toApiEntry("api", apiName)
                .build()
        }

        return this
    }
}

File: client/src/model/ApiSpec.ts
import Hal from "./Hal";
import {OpenAPIV3} from "openapi-types";

export default interface ApiSpec extends Hal {
    apiName: string
    specVersion: string
    status: "ACTIVE" | "HISTORIC" | "PRE_RELEASE" | "ADMIN_ONLY" | "HIDDEN"
    specUrl: string
    spec: any

    // De-referenced document of spec
    document?: OpenAPIV3.Document
}